src/plugins/crypto.c

```diff
@@ -2108,6 +2108,68 @@ gboolean bd_crypto_luks_set_uuid (const gchar *device, const gchar *uuid, GError
     return TRUE;
 }
 
+/**
+ * bd_crypto_luks_convert:
+ * @device:  a LUKS device to convert to a different version of LUKS
+ * @convert_to_luks_version: the LUKS version to convert to
+ * @error: (out) (optional): place to store error (if any)
+ *
+ * Returns: whether the @device was converted to @convert_to_luks_version.
+ *          False, if the @device is already in the @convert_to_luks_version format.
+ *
+ * Warning: LUKS header loss is possible. See bd_crypto_luks_header_backup() and bd_crypto_luks_header_restore()
+ *
+ * Tech category: %BD_CRYPTO_TECH_LUKS-%BD_CRYPTO_TECH_MODE_MODIFY
+ */
+gboolean bd_crypto_luks_convert (const gchar *device, BDCryptoLUKSVersion convert_to_luks_version, GError **error) {
+    struct crypt_device *cd = NULL;
+    gint ret = 0;
+
+    ret = crypt_init (&cd, device);
+    if (ret != 0) {
+        g_set_error (error, BD_CRYPTO_ERROR, BD_CRYPTO_ERROR_DEVICE,
+                     "Failed to initialize device: %s", strerror_l (-ret, c_locale));
+        return FALSE;
+    }
+
+    ret = crypt_load (cd, CRYPT_LUKS, NULL);
+    if (ret != 0) {
+        g_set_error (error, BD_CRYPTO_ERROR, BD_CRYPTO_ERROR_DEVICE,
+                     "Failed to load device: %s", strerror_l (-ret, c_locale));
+        crypt_free (cd);
+        return FALSE;
+    }
+
+    const char *cd_type = crypt_get_type(cd);
+    if (g_strcmp0(cd_type, CRYPT_LUKS1) != 0 && \
+        g_strcmp0 (cd_type, CRYPT_LUKS2) != 0) {
+
+        g_set_error (error, BD_CRYPTO_ERROR, BD_CRYPTO_ERROR_TECH_UNAVAIL,
```

Comment:
Please use `BD_LVM_ERROR_FAIL` here (and in the next error msg too), `BD_LVM_ERROR_TECH_UNAVAIL` is used only in specific cases where support for something is not available due to missing or old dependencies (e.g. cryptsetup version being too old, in this case the operation is simply not possible, not unsupported).

Reply:
I've changed the error to `BD_LVM_ERROR_FAIL` as you asked, but it seems to be a bit off to me -- why LVM if we are in the crypto plugin?

I didn't find explanations to different types of errors; please point me to them if they exist, and I will take them into account next time.

Also, did I specify the tech mode correctly (`BD_CRYPTO_TECH_MODE_MODIFY`)?

Reply:
> I've changed the error to BD_LVM_ERROR_FAIL as you asked, but it seems to be a bit off to me -- why LVM if we are in the crypto plugin?

Yes, using LVM error code is completely wrong here, I have no idea why I recommended this one :man_facepalming: Looking at the error codes we have in the crypto plugin, it looks like we don't have anything that would fit here so please add a new error code, probably something like `BD_CRYPTO_ERROR_CONVERT_FAILED`.

> Also, did I specify the tech mode correctly (BD_CRYPTO_TECH_MODE_MODIFY)?

Yes, that's correct.

Reply:
done, thank you